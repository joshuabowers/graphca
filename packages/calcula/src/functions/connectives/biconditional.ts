import { _ } from '@arrows/multimethod'
import { unit } from "../../monads/writer"
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not } from './complement'
import { and } from './conjunction'
import { implies } from './implication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Biconditional = Connective<Species.xnor>

export const xnorRule = binaryInfixRule(Unicode.xnor)

export const [xnor, isBiconditional, $xnor] = binary<Biconditional, Boolean>(
  Unicode.xnor, Notation.infix, Species.xnor, Genera.connective
)(
  (l, r) => [
    and(implies(unit(l), unit(r)), implies(unit(r), unit(l))), 
    xnorRule(l, r),
    'real biconditional'
  ],
  (l, r) => [
    and(implies(unit(l), unit(r)), implies(unit(r), unit(l))), 
    xnorRule(l, r),
    'complex biconditional'
  ],
  (l, r) => [
    and(implies(unit(l), unit(r)), implies(unit(r), unit(l))), 
    xnorRule(l, r),
    'boolean biconditional'
  ]
)(
  when(
    [isValue(boolean(true)), _],
    (_l, r) => [unit(r), rule`${r}`, 'biconditional identity']
  ),
  when(
    [_, isValue(boolean(true))],
    (l, _r) => [unit(l), rule`${l}`, 'biconditional identity']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [not(unit(r)), rule`${Unicode.not}(${r})`, 'biconditional complementation']
  ),
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [not(unit(l)), rule`${Unicode.not}(${l})`, 'biconditional complementation']
  ),
  when(deepEquals, [boolean(true), rule`${boolean(true)}`, 'biconditional annihilator'])
)
