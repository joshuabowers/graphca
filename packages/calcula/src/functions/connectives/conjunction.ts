import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { isComplement } from './complement'
import { isDisjunction } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Conjunction = Connective<Species.and>

export const andRule = binaryInfixRule(Unicode.and)

export const [and, isConjunction, $and] = binary<Conjunction, Boolean>(
  Species.and, Genera.connective
)(
  (l, r) => [boolean(l.value !== 0 && r.value !== 0), andRule(l, r), 'real conjunction'],
  (l, r) => [
    boolean((l.a !== 0 || l.b !== 0) && (r.a !== 0 || r.b !== 0)), 
    andRule(l, r),
    'complex conjunction'
  ],
  (l, r) => [boolean(l.value && r.value), andRule(l, r), 'boolean conjunction']
)(
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive identity']
  ),
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive identity']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive idempotency']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  )
)
