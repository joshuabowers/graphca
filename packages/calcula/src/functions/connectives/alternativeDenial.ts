import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { TreeNode, Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement, not } from './complement'
import { and } from './conjunction'
import { or } from './disjunction'
import { implies } from './implication'

export type AlternativeDenial = Connective<Species.nand>

export const [nand, isAlternativeDenial, $nand] = binary<AlternativeDenial, Boolean>(
  Species.nand, Genera.connective
)(
  (l, r) => [not(and(unit(l), unit(r))), 'real alternative denial'],
  (l, r) => [not(and(unit(l), unit(r))), 'complex alternative denial'],
  (l, r) => [not(and(unit(l), unit(r))), 'boolean alternative denial']
)(
  when(
    [_, isValue(boolean(true))],
    (l, _r) => [not(unit(l)), 'alternative deniable complementation']
  ),
  when(
    [isValue(boolean(true)), _],
    (_l, r) => [not(unit(r)), 'alternative deniable complementation']
  ),
  when(
    [_, isValue(boolean(false))],
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    [isValue(boolean(false)), _],
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    deepEquals, (l, _r) => [not(unit(l)), 'alternative deniable complementation']
  ),
  when<TreeNode, AlternativeDenial>(
    (l, r) => isAlternativeDenial(r) && deepEquals(l, r.value.left),
    (l, r) => [implies(unit(l), r.right), 'alternative deniable implication']
  ),
  when<TreeNode, AlternativeDenial>(
    (l, r) => isAlternativeDenial(r) && deepEquals(l, r.value.right),
    (l, r) => [implies(unit(l), r.left), 'alternative deniable implication']
  ),
  when<Complement, Complement>(
    (l, r) => isComplement(l) && isComplement(r),
    (l, r) => [or(l.expression, r.expression), 'De Morgan']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'alternative deniable annihilator']
  )
)
