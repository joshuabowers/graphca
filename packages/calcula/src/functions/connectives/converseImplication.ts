import { _ } from '@arrows/multimethod'
import { unit } from "../../monads/writer"
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when, binaryInfixRule } from "../../closures/binary"
import { isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not, notRule } from './complement'
import { or } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type ConverseImplication = Connective<Species.converse>

export const converseRule = binaryInfixRule(Unicode.converse)

export const [converse, isConverseImplication, $converse] = binary<ConverseImplication, Boolean>(
  Unicode.converse, Notation.infix, Species.converse, Genera.connective
)(
  (l, r) => or(unit(l), not(unit(r))),
  (l, r) => or(unit(l), not(unit(r))),
  (l, r) => or(unit(l), not(unit(r)))
)(
  when(
    [isValue(boolean(true)), _], 
    [boolean(true), rule`${boolean(true)}`, 'converse implication annihilator']
  ),
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'converse implication identity']
  ),
  when(
    [isValue(boolean(false)), _], 
    (_l, r) => [not(unit(r)), notRule(r), 'converse implication complementation']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(true), rule`${boolean(true)}`, 'converse implication annihilator']
  )
)
