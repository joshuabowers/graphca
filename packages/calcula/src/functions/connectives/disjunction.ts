import { _ } from '@arrows/multimethod'
import { TreeNode, Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement } from './complement'
import { isConjunction } from './conjunction'
import { implies } from './implication'
import { converse } from './converseImplication'
import { Unicode } from '../../Unicode'

export type Disjunction = Connective<Species.or>

export const [or, isDisjunction, $or] = binary<Disjunction, Boolean>(
  Unicode.or, Notation.infix, Species.or, Genera.connective
)(
  (l, r) => boolean(l.value.value !== 0 || r.value.value !== 0),
  (l, r) => boolean(
    l.value.a !== 0 || l.value.b !== 0 || r.value.a !== 0 || r.value.b !== 0
  ),
  (l, r) => boolean(l.value.value || r.value.value)
)(
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [l, 'disjunctive identity']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [r, 'disjunctive identity']
  ),
  when(
    [_, isValue(boolean(true))],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    [isValue(boolean(true)), _],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [l, 'disjunctive idempotency']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'disjunctive tautology']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'disjunctive tautology']
  ),
  when<Complement, TreeNode>(
    (l, _r) => isComplement(l),
    (l, r) => [
      implies(l.value.expression, r), 
      'disjunctive implication'
    ]
  ),
  when<TreeNode, Complement>(
    (_l, r) => isComplement(r),
    (l, r) => [
      converse(l, r.value.expression), 
      'disjunctive converse implication'
    ]
  )
)
