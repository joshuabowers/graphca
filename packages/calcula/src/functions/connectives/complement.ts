import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation, isSpecies } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { Unary, unary, unaryFnRule, when } from "../../closures/unary"
import { Conjunction, and, andRule } from './conjunction'
import { Disjunction, or, orRule } from './disjunction'
import { ExclusiveDisjunction, xor, xorRule } from './exclusiveDisjunction'
import { Implication } from './implication'
import { AlternativeDenial, nand, nandRule } from './alternativeDenial'
import { JointDenial, nor, norRule } from './jointDenial'
import { Biconditional, xnor, xnorRule } from './biconditional'
import { ConverseImplication } from './converseImplication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Complement = Unary<Species.not, Genera.connective>

export const notRule = unaryFnRule(Unicode.not)

export const [not, isComplement, $not] = unary<Complement, Boolean>(
  Unicode.not, Notation.prefix, Species.not, Genera.connective
)(
  r => [boolean(r.value === 0), notRule(r), 'real complement'],
  c => [boolean(c.a === 0 && c.b === 0), notRule(c), 'complex complement'],
  b => [boolean(!b.value), notRule(b), 'boolean complement']
)(
  // NOTE: Cannot use derived isSpecies guards as they are not defined
  // by this point.
  when<Complement>(
    isSpecies(Species.not), 
    v => [v.expression, rule`${v.expression}`, 'double complement']
  ),
  when<Conjunction>(
    isSpecies(Species.and), 
    v => [nand(v.left, v.right), nandRule(v.left, v.right), 'complement of conjunction']
  ),
  when<Disjunction>(
    isSpecies(Species.or), 
    v => [nor(v.left, v.right), norRule(v.left, v.right), 'complement of disjunction']
  ),
  when<AlternativeDenial>(
    isSpecies(Species.nand), 
    v => [and(v.left, v.right), andRule(v.left, v.right), 'complement of alternative denial']
  ),
  when<JointDenial>(
    isSpecies(Species.nor), 
    v => [or(v.left, v.right), orRule(v.left, v.right), 'complement of joint denial']
  ),
  when<ExclusiveDisjunction>(
    isSpecies(Species.xor), 
    v => [xnor(v.left, v.right), xnorRule(v.left, v.right), 'complement of exclusive disjunction']
  ),
  when<Implication>(
    isSpecies(Species.implies),
    v => [
      and(v.left, not(v.right)), 
      rule`${v.left} ${Unicode.and} (${v.right})${Unicode.not}`, 
      'complement of implication'
    ]
  ),
  when<Biconditional>(
    isSpecies(Species.xnor),
    v => [xor(v.left, v.right), xorRule(v.left, v.right), 'complement of biconditional']
  ),
  when<ConverseImplication>(
    isSpecies(Species.converse),
    v => [
      and(not(v.left), v.right), 
      rule`(${v.left})${Unicode.not} ${Unicode.and} ${v.right}`,
      'complement of converse implication'
    ]
  )
)
