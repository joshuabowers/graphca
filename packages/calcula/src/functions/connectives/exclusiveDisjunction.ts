import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not } from './complement'
import { and } from './conjunction'
import { or } from './disjunction'

export type ExclusiveDisjunction = Connective<Species.xor>

export const [xor, isExclusiveDisjunction, $xor] = binary<ExclusiveDisjunction, Boolean>(
  Species.xor, Genera.connective
)(
  (l, r) => [and(or(unit(l), unit(r)), not(and(unit(l), unit(r)))), 'real exclusive disjunction'],
  (l, r) => [and(or(unit(l), unit(r)), not(and(unit(l), unit(r)))), 'complex exclusive disjunction'],
  (l, r) => [and(or(unit(l), unit(r)), not(and(unit(l), unit(r)))), 'boolean exclusive disjunction']
)(
  // method([bool(false), _], (_l: Boolean, r: Base) => r),
  // method([_, bool(false)], (l: Base, _r: Boolean) => l),
  // method([bool(true), _], (_l: Boolean, r: Base) => not(r)),
  // method([_, bool(true)], (l: Base, _r: Boolean) => not(l)),
  // method(equals, (_l: Base, _r: Base) => bool(false))
)
