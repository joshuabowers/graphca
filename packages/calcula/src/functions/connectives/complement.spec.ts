import { unit } from '../../monads/writer'
import { expectToEqualWithSnapshot } from '../../utility/expectations'
import { Clades, Genera, Species } from '../../utility/tree'
import { real, complex, boolean } from '../../primitives'
import { variable } from '../../variable'
import { not, $not } from './complement'
import { and } from './conjunction'
import { or } from './disjunction'
import { xor } from './exclusiveDisjunction'
import { implies } from './implication'
import { nand } from './alternativeDenial'
import { nor } from './jointDenial'
import { xnor } from './biconditional'
import { converse } from './converseImplication'

describe('$not', () => {
  it('generates a Complement for a TreeNode input', () => {
    expect(
      $not(unit(variable('x').value))[0]
    ).toEqual({
      clade: Clades.unary, genus: Genera.connective, species: Species.not,
      expression: unit(variable('x').value)
    })
  })
})

describe('not', () => {
  it('yields false for non-zero real inputs', () => {
    expectToEqualWithSnapshot(
      not(real(5)),
      boolean(false)
    )
  })

  it('yields true for a real value of zero', () => {
    expectToEqualWithSnapshot(
      not(real(0)),
      boolean(true)
    )
  })

  it('yields false for non-zero complex inputs', () => {
    expectToEqualWithSnapshot(
      not(complex(1, 0)),
      boolean(false)
    )
  })

  it('yields true for complex 0', () => {
    expectToEqualWithSnapshot(
      not(complex(0, 0)),
      boolean(true)
    )
  })

  it('yields false for a true input', () => {
    expectToEqualWithSnapshot(
      not(boolean(true)),
      boolean(false)
    )
  })

  it('yields true for a false input', () => {
    expectToEqualWithSnapshot(
      not(boolean(false)),
      boolean(true)
    )
  })

  it('yields a logical complement for variable input', () => {
    expectToEqualWithSnapshot(
      not(variable('x')),
      $not(variable('x'))[0]
    )
  })

  it('rewrites double negations as the inner expression', () => {
    expectToEqualWithSnapshot(
      not(not(variable('x'))),
      variable('x')
    )
  })

  it('returns an alternative denial when given a conjunction', () => {
    expectToEqualWithSnapshot(
      not(and(variable('x'), variable('y'))),
      nand(variable('x'), variable('y'))
    )
  })

  it('returns a conjunction if given an alternative denial', () => {
    expectToEqualWithSnapshot(
      not(nand(variable('x'), variable('y'))),
      and(variable('x'), variable('y'))
    )
  })

  it('returns a joint denial if given a disjunction', () => {
    expectToEqualWithSnapshot(
      not(or(variable('x'), variable('y'))),
      nor(variable('x'), variable('y'))
    )
  })

  it('returns a disjunction if given a joint denial', () => {
    expectToEqualWithSnapshot(
      not(nor(variable('x'), variable('y'))),
      or(variable('x'), variable('y'))
    )
  })

  it('returns a biconditional if given an exclusive disjunction', () => {
    expectToEqualWithSnapshot(
      not(xor(variable('x'), variable('y'))),
      xnor(variable('x'), variable('y'))
    )
  })

  it('returns a conjunction if given an implication', () => {
    expectToEqualWithSnapshot(
      not(implies(variable('x'), variable('y'))),
      and(variable('x'), not(variable('y')))
    )
  })

  it('returns an exclusive disjunction if given a biconditional', () => {
    expectToEqualWithSnapshot(
      not(xnor(variable('x'), variable('y'))),
      xor(variable('x'), variable('y'))
    )
  })

  it('returns a conjunction if given a converse implication', () => {
    expectToEqualWithSnapshot(
      not(converse(variable('x'), variable('y'))),
      and(not(variable('x')), variable('y'))
    )
  })
})
