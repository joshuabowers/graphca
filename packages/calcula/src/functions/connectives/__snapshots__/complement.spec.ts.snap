// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not returns a biconditional if given an exclusive disjunction 1`] = `
[
  [
    "¬((x⊕y))",
    "identified complement",
  ],
  [
    "¬([(x⊕y)])",
    "processing argument",
  ],
  [
    "(x⊕y)",
    "identified exclusive disjunction",
  ],
  [
    "([x]⊕y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}⊕[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x⊕{y})",
    "processed right operand",
  ],
  [
    "(x⊕y)",
    "created exclusive disjunction",
  ],
  [
    "¬({(x⊕y)})",
    "processed argument",
  ],
  [
    "¬((x⊕y))",
    "complement of exclusive disjunction",
  ],
  [
    "(x↔y)",
    "identified biconditional",
  ],
  [
    "([x]↔y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↔[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↔{y})",
    "processed right operand",
  ],
  [
    "(x↔y)",
    "created biconditional",
  ],
]
`;

exports[`not returns a conjunction if given a converse implication 1`] = `
[
  [
    "¬((x←y))",
    "identified complement",
  ],
  [
    "¬([(x←y)])",
    "processing argument",
  ],
  [
    "(x←y)",
    "identified converse implication",
  ],
  [
    "([x]←y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}←[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x←{y})",
    "processed right operand",
  ],
  [
    "(x←y)",
    "created converse implication",
  ],
  [
    "¬({(x←y)})",
    "processed argument",
  ],
  [
    "¬((x←y))",
    "complement of converse implication",
  ],
  [
    "(¬(x)∧y)",
    "identified conjunction",
  ],
  [
    "([¬(x)]∧y)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(¬(x)∧{y})",
    "processed right operand",
  ],
  [
    "(¬(x)∧y)",
    "created conjunction",
  ],
]
`;

exports[`not returns a conjunction if given an alternative denial 1`] = `
[
  [
    "¬((x↑y))",
    "identified complement",
  ],
  [
    "¬([(x↑y)])",
    "processing argument",
  ],
  [
    "(x↑y)",
    "identified alternative denial",
  ],
  [
    "([x]↑y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↑{y})",
    "processed right operand",
  ],
  [
    "(x↑y)",
    "created alternative denial",
  ],
  [
    "¬({(x↑y)})",
    "processed argument",
  ],
  [
    "¬((x↑y))",
    "complement of alternative denial",
  ],
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
]
`;

exports[`not returns a conjunction if given an implication 1`] = `
[
  [
    "¬((x→y))",
    "identified complement",
  ],
  [
    "¬([(x→y)])",
    "processing argument",
  ],
  [
    "(x→y)",
    "identified implication",
  ],
  [
    "([x]→y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}→[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x→{y})",
    "processed right operand",
  ],
  [
    "(x→y)",
    "created implication",
  ],
  [
    "¬({(x→y)})",
    "processed argument",
  ],
  [
    "¬((x→y))",
    "complement of implication",
  ],
  [
    "(x∧¬(y))",
    "identified conjunction",
  ],
  [
    "([x]∧¬(y))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[¬(y)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(y)",
    "identified complement",
  ],
  [
    "¬([y])",
    "processing argument",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "¬({y})",
    "processed argument",
  ],
  [
    "¬(y)",
    "created complement",
  ],
  [
    "(x∧{¬(y)})",
    "processed right operand",
  ],
  [
    "(x∧¬(y))",
    "created conjunction",
  ],
]
`;

exports[`not returns a disjunction if given a joint denial 1`] = `
[
  [
    "¬((x↓y))",
    "identified complement",
  ],
  [
    "¬([(x↓y)])",
    "processing argument",
  ],
  [
    "(x↓y)",
    "identified joint denial",
  ],
  [
    "([x]↓y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↓{y})",
    "processed right operand",
  ],
  [
    "(x↓y)",
    "created joint denial",
  ],
  [
    "¬({(x↓y)})",
    "processed argument",
  ],
  [
    "¬((x↓y))",
    "complement of joint denial",
  ],
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
]
`;

exports[`not returns a joint denial if given a disjunction 1`] = `
[
  [
    "¬((x∨y))",
    "identified complement",
  ],
  [
    "¬([(x∨y)])",
    "processing argument",
  ],
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
  [
    "¬({(x∨y)})",
    "processed argument",
  ],
  [
    "¬((x∨y))",
    "complement of disjunction",
  ],
  [
    "(x↓y)",
    "identified joint denial",
  ],
  [
    "([x]↓y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↓{y})",
    "processed right operand",
  ],
  [
    "(x↓y)",
    "created joint denial",
  ],
]
`;

exports[`not returns an alternative denial when given a conjunction 1`] = `
[
  [
    "¬((x∧y))",
    "identified complement",
  ],
  [
    "¬([(x∧y)])",
    "processing argument",
  ],
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
  [
    "¬({(x∧y)})",
    "processed argument",
  ],
  [
    "¬((x∧y))",
    "complement of conjunction",
  ],
  [
    "(x↑y)",
    "identified alternative denial",
  ],
  [
    "([x]↑y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↑{y})",
    "processed right operand",
  ],
  [
    "(x↑y)",
    "created alternative denial",
  ],
]
`;

exports[`not returns an exclusive disjunction if given a biconditional 1`] = `
[
  [
    "¬((x↔y))",
    "identified complement",
  ],
  [
    "¬([(x↔y)])",
    "processing argument",
  ],
  [
    "(x↔y)",
    "identified biconditional",
  ],
  [
    "([x]↔y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↔[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↔{y})",
    "processed right operand",
  ],
  [
    "(x↔y)",
    "created biconditional",
  ],
  [
    "¬({(x↔y)})",
    "processed argument",
  ],
  [
    "¬((x↔y))",
    "complement of biconditional",
  ],
  [
    "(x⊕y)",
    "identified exclusive disjunction",
  ],
  [
    "([x]⊕y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}⊕[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x⊕{y})",
    "processed right operand",
  ],
  [
    "(x⊕y)",
    "created exclusive disjunction",
  ],
]
`;

exports[`not rewrites double negations as the inner expression 1`] = `
[
  [
    "¬(¬(x))",
    "identified complement",
  ],
  [
    "¬([¬(x)])",
    "processing argument",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "¬({¬(x)})",
    "processed argument",
  ],
  [
    "¬(¬(x))",
    "double complement",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`not yields a logical complement for variable input 1`] = `
[
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`not yields false for a true input 1`] = `
[
  [
    "¬(true)",
    "identified complement",
  ],
  [
    "¬([true])",
    "processing argument",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "¬({true})",
    "processed argument",
  ],
  [
    "¬(true)",
    "boolean complement",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`not yields false for non-zero complex inputs 1`] = `
[
  [
    "¬(1+0𝒊)",
    "identified complement",
  ],
  [
    "¬([1+0𝒊])",
    "processing argument",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "¬({1+0𝒊})",
    "processed argument",
  ],
  [
    "¬(1+0𝒊)",
    "complex complement",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`not yields false for non-zero real inputs 1`] = `
[
  [
    "¬(5)",
    "identified complement",
  ],
  [
    "¬([5])",
    "processing argument",
  ],
  [
    "5",
    "created real",
  ],
  [
    "¬({5})",
    "processed argument",
  ],
  [
    "¬(5)",
    "real complement",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`not yields true for a false input 1`] = `
[
  [
    "¬(false)",
    "identified complement",
  ],
  [
    "¬([false])",
    "processing argument",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "¬({false})",
    "processed argument",
  ],
  [
    "¬(false)",
    "boolean complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`not yields true for a real value of zero 1`] = `
[
  [
    "¬(0)",
    "identified complement",
  ],
  [
    "¬([0])",
    "processing argument",
  ],
  [
    "0",
    "created real",
  ],
  [
    "¬({0})",
    "processed argument",
  ],
  [
    "¬(0)",
    "real complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`not yields true for complex 0 1`] = `
[
  [
    "¬(0+0𝒊)",
    "identified complement",
  ],
  [
    "¬([0+0𝒊])",
    "processing argument",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "¬({0+0𝒊})",
    "processed argument",
  ],
  [
    "¬(0+0𝒊)",
    "complex complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
